client/my-sites/controller.js

```diff
@@ -109,6 +112,43 @@ function renderNoVisibleSites( context ) {
 	);
 }
 
+function isPathAllowedForDomainOnlySite( pathname ) {
+	const urlPrefixesWhiteListForDomainOnlySite = [
+		'/domains/manage/',
+		'/checkout/',
+	];
+
+	return urlPrefixesWhiteListForDomainOnlySite.some( path => startsWith( pathname, path ) );
+}
+
+function onSelectedSiteAvailable( context ) {
+	const selectedSite = sites.getSelectedSite();
+	siteStatsStickyTabActions.saveFilterAndSlug( false, selectedSite.slug );
+	context.store.dispatch( receiveSite( selectedSite ) );
+	context.store.dispatch( setSelectedSiteId( selectedSite.ID ) );
+
+	const state = context.store.getState();
+
+	if ( isDomainOnlySite( state, selectedSite.ID ) &&
+		! isPathAllowedForDomainOnlySite( context.pathname ) ) {
+		page.redirect( domainManagementList( selectedSite.slug ) );
+		return false;
+	}
+
+	// Update recent sites preference
+	if ( hasReceivedRemotePreferences( state ) ) {
+		const recentSites = getPreference( state, 'recentSites' );
+		if ( selectedSite.ID !== recentSites[ 0 ] ) {
+			context.store.dispatch( savePreference( 'recentSites', uniq( [
+				selectedSite.ID,
+				...recentSites
+			] ).slice( 0, 3 ) ) );
```

Comment:
Yeah, it would be great if it’s in a constant, [magic numbers](http://wiki.c2.com/?MagicNumber) can be confusing.

Also, do you think we should so much preprocessing before dispatch?